File: examples/dotnet/producer/Scenarios.cs
using DaprMQ.Client;

/// <summary>Producer side of the 4 demo scenarios - see examples/shared/SCENARIOS.md.</summary>
public static class Scenarios
{
    public static readonly IReadOnlyList<ScenarioDescriptor> Descriptors = new List<ScenarioDescriptor>
    {
        new("basic", "producer", "Enqueues 3 plain items."),
        new("ack-deadletter", "producer", "Enqueues 3 items tagged ack/deadletter/expire."),
        new("priority", "producer", "Enqueues normal-priority items, then fast-lane items."),
        new("sessions", "producer", "Enqueues items across two sessions."),
        new("idempotency", "producer", "Enqueues an item, a duplicate reusing its idempotency key, then a distinct item."),
    };

    public static bool IsKnown(string name) => Descriptors.Any(d => d.Name == name);

    public static Task<(string QueueId, List<StepDto> Steps)> RunAsync(
        string name, IDaprMQClient client, string queuePrefix, Action<string, string> log) => name switch
    {
        "basic" => RunBasicAsync(client, queuePrefix, log),
        "ack-deadletter" => RunAckDeadletterAsync(client, queuePrefix, log),
        "priority" => RunPriorityAsync(client, queuePrefix, log),
        "sessions" => RunSessionsAsync(client, queuePrefix, log),
        "idempotency" => RunIdempotencyAsync(client, queuePrefix, log),
        _ => throw new ArgumentOutOfRangeException(nameof(name), $"unknown scenario '{name}'")
    };

    private static void Step(List<StepDto> steps, Action<string, string> log, string action, string detail)
    {
        log("INFO", detail);
        steps.Add(new StepDto(action, detail));
    }

    private static async Task<(string, List<StepDto>)> RunBasicAsync(
        IDaprMQClient client, string queuePrefix, Action<string, string> log)
    {
        var queueId = $"{queuePrefix}-basic";
        var steps = new List<StepDto>();

        var items = new[]
        {
            new EnqueueItemDto(new { n = 1, message = "hello from producer" }),
            new EnqueueItemDto(new { n = 2, message = "hello from producer" }),
            new EnqueueItemDto(new { n = 3, message = "hello from producer" }),
        };

        await client.EnqueueAsync(queueId, items);

        for (var i = 0; i < items.Length; i++)
        {
            Step(steps, log, "enqueue", $"Enqueued item {i + 1}/{items.Length} to queue {queueId} (priority={items[i].Priority})");
        }

        return (queueId, steps);
    }

    private static async Task<(string, List<StepDto>)> RunAckDeadletterAsync(
        IDaprMQClient client, string queuePrefix, Action<string, string> log)
    {
        var queueId = $"{queuePrefix}-ackdlq";
        var steps = new List<StepDto>();

        var payloads = new (string Outcome, int N)[]
        {
            ("ack", 1),
            ("deadletter", 2),
            ("expire", 3),
        };
        var items = payloads.Select(p => new EnqueueItemDto(new { outcome = p.Outcome, n = p.N })).ToArray();

        await client.EnqueueAsync(queueId, items);

        foreach (var p in payloads)
        {
            Step(steps, log, "enqueue", $"Enqueued item to queue {queueId}: outcome={p.Outcome}, n={p.N}");
        }

        return (queueId, steps);
    }

    private static async Task<(string, List<StepDto>)> RunPriorityAsync(
        IDaprMQClient client, string queuePrefix, Action<string, string> log)
    {
        var queueId = $"{queuePrefix}-priority";
        var steps = new List<StepDto>();

        // Deliberately priority-inverted enqueue order: normal (priority=1) items first,
        // then fast-lane (priority=0) items, to demonstrate priority ordering on dequeue.
        var normal = new[]
        {
            new EnqueueItemDto(new { priority = 1, n = 1 }, Priority: 1),
            new EnqueueItemDto(new { priority = 1, n = 2 }, Priority: 1),
            new EnqueueItemDto(new { priority = 1, n = 3 }, Priority: 1),
        };
        await client.EnqueueAsync(queueId, normal);
        for (var i = 0; i < normal.Length; i++)
        {
            Step(steps, log, "enqueue", $"Enqueued item {i + 1}/3 (n={i + 1}) to queue {queueId} (priority=1)");
        }

        var fast = new[]
        {
            new EnqueueItemDto(new { priority = 0, n = 4 }, Priority: 0),
            new EnqueueItemDto(new { priority = 0, n = 5 }, Priority: 0),
            new EnqueueItemDto(new { priority = 0, n = 6 }, Priority: 0),
        };
        await client.EnqueueAsync(queueId, fast);
        for (var i = 0; i < fast.Length; i++)
        {
            Step(steps, log, "enqueue", $"Enqueued item {i + 1}/3 (n={i + 4}) to queue {queueId} (priority=0)");
        }

        return (queueId, steps);
    }

    private static async Task<(string, List<StepDto>)> RunSessionsAsync(
        IDaprMQClient client, string queuePrefix, Action<string, string> log)
    {
        var queueId = $"{queuePrefix}-sessions";
        var steps = new List<StepDto>();

        var items = new[]
        {
            new EnqueueItemDto(new { sessionId = "session-a", seq = 1 }, SessionId: "session-a"),
            new EnqueueItemDto(new { sessionId = "session-a", seq = 2 }, SessionId: "session-a"),
            new EnqueueItemDto(new { sessionId = "session-b", seq = 1 }, SessionId: "session-b"),
            new EnqueueItemDto(new { sessionId = "session-b", seq = 2 }, SessionId: "session-b"),
        };

        await client.EnqueueAsync(queueId, items);

        foreach (var item in items)
        {
            Step(steps, log, "enqueue", $"Enqueued item to queue {queueId} for session {item.SessionId}");
        }

        return (queueId, steps);
    }

    /// <summary>
    /// Enqueues an item, a duplicate reusing its idempotency key (expected to be
    /// deduplicated), then a distinct item under a different key. The key is suffixed
    /// with the current time so repeated runs don't collide with a previous run's key
    /// still inside the dedup TTL window.
    /// </summary>
    private static async Task<(string, List<StepDto>)> RunIdempotencyAsync(
        IDaprMQClient client, string queuePrefix, Action<string, string> log)
    {
        var queueId = $"{queuePrefix}-idempotency";
        var steps = new List<StepDto>();
        var key = $"idempotency-demo-{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}";

        var first = await client.EnqueueAsync(queueId, new[] { new EnqueueItemDto(new { n = 1 }, IdempotencyKey: key) });
        Step(steps, log, "enqueue",
            $"Enqueued item n=1 with idempotencyKey={key} (itemsEnqueued={first.ItemsEnqueued}, itemsDeduplicated={first.ItemsDeduplicated})");

        var duplicate = await client.EnqueueAsync(queueId, new[] { new EnqueueItemDto(new { n = 2 }, IdempotencyKey: key) });
        Step(steps, log, "enqueue",
            $"Enqueued item n=2 reusing idempotencyKey={key} (itemsEnqueued={duplicate.ItemsEnqueued}, itemsDeduplicated={duplicate.ItemsDeduplicated}) - expected to be deduplicated");

        var distinctKey = $"{key}-b";
        var distinct = await client.EnqueueAsync(queueId, new[] { new EnqueueItemDto(new { n = 3 }, IdempotencyKey: distinctKey) });
        Step(steps, log, "enqueue",
            $"Enqueued item n=3 with a different idempotencyKey={distinctKey} (itemsEnqueued={distinct.ItemsEnqueued}, itemsDeduplicated={distinct.ItemsDeduplicated})");

        return (queueId, steps);
    }
}

File: examples/typescript/producer/src/scenarios.ts
import { getClient, getQueuePrefix } from "./config.js";
import { log } from "./logger.js";

export interface ScenarioStep {
  action: string;
  detail: string;
}

export interface ScenarioRunResult {
  queueId: string;
  steps: ScenarioStep[];
}

export type ScenarioName = "basic" | "ack-deadletter" | "priority" | "sessions" | "idempotency";

export interface ScenarioDescriptor {
  name: ScenarioName;
  role: "producer";
  description: string;
  run: () => Promise<ScenarioRunResult>;
}

/** basic - enqueues 3 plain items in a single Enqueue call, default priority, no session. */
async function runBasic(): Promise<ScenarioRunResult> {
  const queueId = `${getQueuePrefix()}-basic`;
  const client = getClient();

  const items = [1, 2, 3].map((n) => ({ item: { n, message: "hello from producer" } }));
  await client.enqueue(queueId, items);

  const steps: ScenarioStep[] = [];
  items.forEach((_entry, index) => {
    const detail = `Enqueued item ${index + 1}/${items.length} to queue ${queueId} (priority=1)`;
    log("INFO", detail);
    steps.push({ action: "enqueue", detail });
  });

  return { queueId, steps };
}

/** ack-deadletter - enqueues 3 items tagged with the outcome the consumer should apply. */
async function runAckDeadletter(): Promise<ScenarioRunResult> {
  const queueId = `${getQueuePrefix()}-ackdlq`;
  const client = getClient();

  const items = [
    { item: { outcome: "ack", n: 1 } },
    { item: { outcome: "deadletter", n: 2 } },
    { item: { outcome: "expire", n: 3 } },
  ];
  await client.enqueue(queueId, items);

  const steps: ScenarioStep[] = [];
  items.forEach(({ item }) => {
    const { outcome, n } = item as { outcome: string; n: number };
    const detail = `Enqueued item ${n}/${items.length} (outcome=${outcome}) to queue ${queueId}`;
    log("INFO", detail);
    steps.push({ action: "enqueue", detail });
  });

  return { queueId, steps };
}

/**
 * priority - enqueues 3 normal-priority (1) items first, then 3 fast-lane (0) items, deliberately
 * priority-inverted so the consumer scenario can demonstrate the fast-lane items surfacing first.
 */
async function runPriority(): Promise<ScenarioRunResult> {
  const queueId = `${getQueuePrefix()}-priority`;
  const client = getClient();
  const steps: ScenarioStep[] = [];

  const normalItems = [1, 2, 3].map((n) => ({ item: { priority: 1, n }, priority: 1 }));
  await client.enqueue(queueId, normalItems);
  normalItems.forEach(({ item }) => {
    const { n } = item as { n: number };
    const detail = `Enqueued item n=${n} to queue ${queueId} (priority=1, normal)`;
    log("INFO", detail);
    steps.push({ action: "enqueue", detail });
  });

  const fastLaneItems = [4, 5, 6].map((n) => ({ item: { priority: 0, n }, priority: 0 }));
  await client.enqueue(queueId, fastLaneItems);
  fastLaneItems.forEach(({ item }) => {
    const { n } = item as { n: number };
    const detail = `Enqueued item n=${n} to queue ${queueId} (priority=0, fast-lane)`;
    log("INFO", detail);
    steps.push({ action: "enqueue", detail });
  });

  return { queueId, steps };
}

/** sessions - enqueues 4 items, 2 per session, each carrying a per-session sequence number. */
async function runSessions(): Promise<ScenarioRunResult> {
  const queueId = `${getQueuePrefix()}-sessions`;
  const client = getClient();

  const items = [
    { item: { sessionId: "session-a", seq: 1 }, sessionId: "session-a" },
    { item: { sessionId: "session-a", seq: 2 }, sessionId: "session-a" },
    { item: { sessionId: "session-b", seq: 1 }, sessionId: "session-b" },
    { item: { sessionId: "session-b", seq: 2 }, sessionId: "session-b" },
  ];
  await client.enqueue(queueId, items);

  const steps: ScenarioStep[] = [];
  items.forEach((entry) => {
    const { sessionId, seq } = entry.item as { sessionId: string; seq: number };
    const detail = `Enqueued item seq=${seq} for session=${sessionId} to queue ${queueId}`;
    log("INFO", detail);
    steps.push({ action: "enqueue", detail });
  });

  return { queueId, steps };
}

/**
 * idempotency - enqueues one item with a fresh idempotencyKey, then a second item reusing the
 * SAME key (expected to be silently deduplicated), then a third item with a different key
 * (expected to enqueue normally). The key is suffixed with the current time so repeated runs of
 * this scenario don't collide with a previous run's key still inside the dedup TTL window.
 */
async function runIdempotency(): Promise<ScenarioRunResult> {
  const queueId = `${getQueuePrefix()}-idempotency`;
  const client = getClient();
  const steps: ScenarioStep[] = [];
  const key = `idempotency-demo-${Date.now()}`;

  const first = await client.enqueue(queueId, [{ item: { n: 1 }, idempotencyKey: key }]);
  let detail = `Enqueued item n=1 with idempotencyKey=${key} (itemsEnqueued=${first.itemsEnqueued}, itemsDeduplicated=${first.itemsDeduplicated})`;
  log("INFO", detail);
  steps.push({ action: "enqueue", detail });

  const duplicate = await client.enqueue(queueId, [{ item: { n: 2 }, idempotencyKey: key }]);
  detail = `Enqueued item n=2 reusing idempotencyKey=${key} (itemsEnqueued=${duplicate.itemsEnqueued}, itemsDeduplicated=${duplicate.itemsDeduplicated}) - expected to be deduplicated`;
  log("INFO", detail);
  steps.push({ action: "enqueue", detail });

  const distinctKey = `${key}-b`;
  const distinct = await client.enqueue(queueId, [{ item: { n: 3 }, idempotencyKey: distinctKey }]);
  detail = `Enqueued item n=3 with a different idempotencyKey=${distinctKey} (itemsEnqueued=${distinct.itemsEnqueued}, itemsDeduplicated=${distinct.itemsDeduplicated})`;
  log("INFO", detail);
  steps.push({ action: "enqueue", detail });

  return { queueId, steps };
}

export const SCENARIOS: ScenarioDescriptor[] = [
  { name: "basic", role: "producer", description: "Enqueues 3 plain items.", run: runBasic },
  {
    name: "ack-deadletter",
    role: "producer",
    description: "Enqueues 3 items tagged ack/deadletter/expire.",
    run: runAckDeadletter,
  },
  {
    name: "priority",
    role: "producer",
    description: "Enqueues normal-priority items, then fast-lane items.",
    run: runPriority,
  },
  { name: "sessions", role: "producer", description: "Enqueues items across two sessions.", run: runSessions },
  {
    name: "idempotency",
    role: "producer",
    description: "Enqueues an item, a duplicate reusing its idempotency key, then a distinct item.",
    run: runIdempotency,
  },
];
